runtime/interpreter/value.go

```diff
@@ -921,122 +992,126 @@ func (v *ArrayValue) Insert(inter *Interpreter, getLocationRange func() Location
 		})
 	}
 
-	// Use the getter, to make sure lazily decoded values are properly loaded.
-	arrayStaticType := v.StaticType().(ArrayStaticType)
-
-	inter.checkContainerMutation(arrayStaticType.ElementType(), element, getLocationRange)
+	interpreter.checkContainerMutation(v.Type.ElementType(), element, getLocationRange)
 
-	v.modified = true
-
-	element.SetOwner(v.Owner)
-
-	elements := v.Elements()
-
-	//nolint:gocritic
-	v.values = append(
-		elements[:index],
-		append(
-			[]Value{element},
-			elements[index:]...,
-		)...,
+	element = interpreter.TransferValue(
+		getLocationRange,
+		element,
+		nil,
+		v.array.Address(),
 	)
-}
-
-// TODO: unset owner?
-func (v *ArrayValue) Remove(index int, getLocationRange func() LocationRange) Value {
-	v.checkBounds(index, getLocationRange)
-
-	v.modified = true
 
-	elements := v.Elements()
-	result := elements[index]
-
-	lastIndex := len(elements) - 1
-	copy(elements[index:], elements[index+1:])
-
-	// avoid memory leaks by explicitly setting value to nil
-	elements[lastIndex] = nil
-
-	v.values = elements[:lastIndex]
+	err := v.array.Insert(uint64(index), element)
+	if err != nil {
+		panic(ExternalError{err})
+	}
+	interpreter.maybeCheckAtreeValue(v.array)
+}
 
-	return result
+func (v *ArrayValue) RemoveKey(interpreter *Interpreter, getLocationRange func() LocationRange, key Value) Value {
+	index := key.(NumberValue).ToInt()
+	return v.Remove(interpreter, getLocationRange, index)
 }
 
-// TODO: unset owner?
-func (v *ArrayValue) RemoveFirst(getLocationRange func() LocationRange) Value {
-	v.checkBounds(0, getLocationRange)
+func (v *ArrayValue) Remove(interpreter *Interpreter, getLocationRange func() LocationRange, index int) Value {
+	storable, err := v.array.Remove(uint64(index))
+	if err != nil {
+		if _, ok := err.(*atree.IndexOutOfBoundsError); ok {
+			panic(ArrayIndexOutOfBoundsError{
+				Index:         index,
+				Size:          v.Count(),
+				LocationRange: getLocationRange(),
+			})
+		}
 
-	v.modified = true
+		panic(ExternalError{err})
+	}
+	interpreter.maybeCheckAtreeValue(v.array)
 
-	elements := v.Elements()
+	value := StoredValue(storable, interpreter.Storage)
 
-	var firstElement Value
-	firstElement, v.values = elements[0], elements[1:]
-	return firstElement
+	return interpreter.TransferValue(
+		getLocationRange,
+		value,
+		storable,
+		atree.Address{},
```

Comment:
What do you think about making `atree.Address{}` a global const, like `const emptyAddress = atree.Address{}`?